File: astrx-suite/crawlcore/crawlcore/globmatch.py
"""Linear (backtracking-free) robots.txt path-glob matcher, shared by both
crawlers' robots parsers so there is ONE tested, ReDoS-safe implementation.

A robots path pattern uses ``*`` as a wildcard (any run of characters) and a
trailing ``$`` as an end-anchor; matching is ALWAYS anchored at the start of the
path (the de-facto robots rule). This is implemented with ``str`` prefix / find /
endswith scans only -- there is no regex and therefore no catastrophic
backtracking, so a hostile robots.txt whose Disallow pattern is (e.g.)
``/a*a*a*...*$`` can never hang the crawl (ReDoS).

Semantics are identical to ``re.match`` of the translated pattern
(``*`` -> ``.*``, optional trailing ``$``), which is what both crawlers relied
on before this extraction:

  * no wildcard, unanchored -> ``path.startswith(pattern)``
  * no wildcard, anchored   -> ``path == pattern``
  * wildcards               -> the literal segments must appear in order, the
                               first anchored at the start of the path, and (when
                               ``$``) the last segment must sit at the very end.
"""

from __future__ import annotations

import re

# Bound the pattern length as belt-and-suspenders. The matcher is linear
# regardless, but truncating keeps even pathological input trivially cheap.
# Truncating a literal only makes a Disallow prefix SHORTER (matching more, i.e.
# fetching less), so it can never turn a Disallow into an accidental over-fetch.
MAX_PATTERN_LEN = 4096

_STAR_RUN = re.compile(r"\*+")


def compile_glob(pattern: str):
    """Split a robots path pattern into ``(anchored_end, literal_segments)``.

    ``*`` is a wildcard, a trailing ``$`` anchors to the end of the path. Runs of
    ``*`` are collapsed (``.*.*`` == ``.*``) so the segment list stays minimal.
    """
    if len(pattern) > MAX_PATTERN_LEN:
        pattern = pattern[:MAX_PATTERN_LEN]
    anchored = pattern.endswith("$")
    if anchored:
        pattern = pattern[:-1]
    pattern = _STAR_RUN.sub("*", pattern)   # collapse runs of '*' (linear)
    return anchored, pattern.split("*")


def glob_match(segments, anchored: bool, path: str) -> bool:
    """Linear wildcard match of *path* against pre-split robots *segments*.

    Mirrors ``re.match`` of the translated pattern: anchored at the start of
    *path*, wildcards between literal segments, optional end-anchor.
    """
    n = len(segments)
    if n == 1:                       # no wildcard
        seg = segments[0]
        return path == seg if anchored else path.startswith(seg)
    if not path.startswith(segments[0]):
        return False
    pos = len(segments[0])
    for seg in segments[1:-1]:
        if not seg:
            continue
        idx = path.find(seg, pos)
        if idx == -1:
            return False
        pos = idx + len(seg)
    last = segments[-1]
    if not last:                     # pattern ended with '*' (or '*$')
        return True
    if anchored:
        return path.endswith(last) and len(path) - len(last) >= pos
    return path.find(last, pos) != -1

File: rewrite/crawlcore-go/globmatch.go
// Package crawlcore is the Go port of the shared crawl library.
//
// This file is the backtracking-free, ReDoS-safe robots.txt path-glob matcher
// (a port of crawlcore/globmatch.py). A robots path pattern uses '*' as a
// wildcard and a trailing '$' as an end-anchor; matching is always anchored at
// the start of the path. It is implemented with prefix/index/suffix scans only —
// no regex, so a hostile Disallow pattern like "/a*a*a*...*$" can never hang the
// crawl. Semantics are identical to re.match of the translated pattern
// ('*' -> '.*', optional trailing '$', start-anchored).
package crawlcore

import "strings"

// MaxPatternLen bounds the pattern length as belt-and-suspenders. The matcher is
// linear regardless, but truncating keeps even pathological input trivially
// cheap. Truncating a literal only makes a Disallow prefix shorter (matching
// more, i.e. fetching less), so it can never become an accidental over-fetch.
const MaxPatternLen = 4096

// CompileGlob splits a robots path pattern into its end-anchor flag and the
// literal segments between '*' wildcards. Runs of '*' collapse to one so the
// segment list stays minimal.
func CompileGlob(pattern string) (anchored bool, segments []string) {
	if len(pattern) > MaxPatternLen {
		pattern = pattern[:MaxPatternLen]
	}
	anchored = strings.HasSuffix(pattern, "$")
	if anchored {
		pattern = pattern[:len(pattern)-1]
	}
	return anchored, strings.Split(collapseStars(pattern), "*")
}

// collapseStars replaces every run of '*' with a single '*' (linear, no regex).
func collapseStars(s string) string {
	if !strings.Contains(s, "**") {
		return s
	}
	var b strings.Builder
	b.Grow(len(s))
	prevStar := false
	for i := 0; i < len(s); i++ {
		if s[i] == '*' {
			if prevStar {
				continue
			}
			prevStar = true
		} else {
			prevStar = false
		}
		b.WriteByte(s[i])
	}
	return b.String()
}

// GlobMatch reports whether path matches the pre-split robots segments: anchored
// at the start of path, wildcards between literal segments, optional end-anchor.
func GlobMatch(segments []string, anchored bool, path string) bool {
	n := len(segments)
	if n == 1 { // no wildcard
		seg := segments[0]
		if anchored {
			return path == seg
		}
		return strings.HasPrefix(path, seg)
	}
	if !strings.HasPrefix(path, segments[0]) {
		return false
	}
	pos := len(segments[0])
	for _, seg := range segments[1 : n-1] {
		if seg == "" {
			continue
		}
		idx := strings.Index(path[pos:], seg)
		if idx == -1 {
			return false
		}
		pos += idx + len(seg)
	}
	last := segments[n-1]
	if last == "" { // pattern ended with '*' (or '*$')
		return true
	}
	if anchored {
		return strings.HasSuffix(path, last) && len(path)-len(last) >= pos
	}
	return strings.Index(path[pos:], last) != -1
}
